osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -558,71 +573,98 @@ protected override void Update()
-                            break;
+                    panel.Depth = item.CarouselYPosition;
+                    panel.Y = item.CarouselYPosition;
 
-                        case LoadState.Loading:
-                            break;
+                    ScrollableContent.Add(panel);
+                }
+            }
 
-                        default:
-                            scrollableContent.Add(item);
-                            break;
-                    }
+            // Finally, if the filtered items have changed, animate drawables to their new locations.
+            // This is common if a selected/collapsed state has changed.
+            if (revalidateItems)
+            {
+                foreach (DrawableCarouselItem panel in ScrollableContent.Children)
+                {
+                    panel.MoveToY(panel.Item.CarouselYPosition, 800, Easing.OutQuint);
                 }
-                else
+            }
+
+            // Update externally controlled state of currently visible items (e.g. x-offset and opacity).
+            // This is a per-frame update on all drawable panels.
+            foreach (DrawableCarouselItem item in ScrollableContent.Children)
+            {
+                updateItem(item);
+
+                if (item is DrawableCarouselBeatmapSet set)
                 {
-                    scrollableContent.ChangeChildDepth(item, depth);
+                    foreach (var diff in set.DrawableBeatmaps)
+                        updateItem(diff, item);
                 }
             }
+        }
 
-            // this is not actually useful right now, but once we have groups may well be.
-            if (notVisibleCount > 50)
-                itemsCache.Invalidate();
+        private readonly CarouselBoundsItem carouselBoundsItem = new CarouselBoundsItem();
+
+        private (int firstIndex, int lastIndex) getDisplayRange()
+        {
+            // Find index range of all items that should be on-screen
+            carouselBoundsItem.CarouselYPosition = visibleUpperBound - distance_offscreen_to_preload;
+            int firstIndex = visibleItems.BinarySearch(carouselBoundsItem);
+            if (firstIndex < 0) firstIndex = ~firstIndex;
 
-            // Update externally controlled state of currently visible items
-            // (e.g. x-offset and opacity).
-            foreach (DrawableCarouselItem p in scrollableContent.Children)
-                updateItem(p);
+            carouselBoundsItem.CarouselYPosition = visibleBottomBound + distance_offscreen_to_preload;
+            int lastIndex = visibleItems.BinarySearch(carouselBoundsItem);
+            if (lastIndex < 0) lastIndex = ~lastIndex;
+
+            // as we can't be 100% sure on the size of individual carousel drawables,
+            // always play it safe and extend bounds by one.
+            firstIndex = Math.Max(0, firstIndex - 1);
+            lastIndex = Math.Min(visibleItems.Count, lastIndex + 1);
```

Comment:
Crashes when scrolling to the end of the list. Probably needs to be `visibleItems.Count - 1`? It should also probably be clamped to `firstIndex`